docs/_docs/reference/other-new-features/safe-initialization.md

```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
-   has different syntax for them, it thus is not an issue.
+   initialization `val f: T = e`, the expression `e` may point to an non-hot
+   value.
 
-2. Objects under initialization may not be passed as arguments to method calls.
+4. Arguments to method calls must be _effectively hot_.
 
    Escape of `this` in the constructor is commonly regarded as an anti-pattern.
+
    However, escape of `this` as constructor arguments are allowed, to support
    creation of cyclic data structures. The checker will ensure that the escaped
    non-initialized object is not used, i.e. calling methods or accessing fields
    on the escaped object is not allowed.
 
+   An exception is for calling synthetic `apply`s of case classes. For example,
+   the method call `Some.apply(e)` will be interpreted as `new Some(e)`, thus
+   is valid even if `e` is not hot.
+
+   Another exception too this rule is parametric method calls. For example, in
+   `List.apply(e)`, the argument `e` may be non-hot. If that is the case, the
+   result value of the parametric method call is taken as _cold_.
+
+5. Method calls on hot values with effectively hot arguments produce hot results.
+
+   This rule is assured by local reasoning about initialization.
+
+6. Method calls on `ThisRef` and warm values will be resolved statically and the
+   corresponding method bodies are checked.
+
+7. In a new expression `new p.C(args)`, if the values of `p` and `args` are
+   effectively hot, then the result value is also hot.
+
+   This is assured by local reasoning about initialization.
+
+8. In a new expression `new p.C(args)`, if any value of `p` and `args` is not
+   effectively hot, then the result value takes the form `Warm[C] { outer = Vp, args = Vargs }`. The initialization code for the class `C` is checked again to make
+   sure the non-hot values are used properly.
+
+   In the above, `Vp` is the widened value of `p` --- the widening happens if `p`
+   is a warm value `Warm[D] { outer = V, args }` and we widen it to
+   `Warm[D] { outer = Cold, args }`.
+
+   The variable `Vargs` represents values of `args` with non-hot values widened
+   to `Cold`.
+
+   The widening is motivated to finitize the abstract domain and ensure
+   terimination of the initialization check.
+
+9. The scrutinee in pattern match, the values in return and throw statements must be _effectively hot_.
+
+A value `v` is _effectively hot_ if any of the following is true:
+
+- `v` is `Hot`.
+- `v` is `ThisRef` and all fields of the underlying object are initialized.
+- `v` is `Warm[C] { outer = V, ctor, args = Vs }` and
+  1. `C` does not contain inner classes;
+  2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_;
+  3. Each field of `v` is _effectively hot_.
+- `v` is `Fun(e, V, C)` and calling the function encounters no errors and the
+  function return value is _effectively hot_.
+- `ThisRef` is _effectively hot_.
+
```

Comment:
```suggestion
In the `ThisRef` case, local reasoning ensures that all fields assigned implies that all fields themselves contain effectively hot values.

In the `Warm` case, if the value `v` leaks outside the analysis scope, the unanalyzed code could instantiate inner classes of `v`, call methods on `v`, or read fields of `v`. The conditions of the `Warm` case ensure that these actions do not lead to accessing a cold object.

In the `Fun` case, ...
```

Reply:
Maybe we can add the justification later --- they are not immediately understandable to readers and can itself be an obstacle in reading the docs.